teslajsonpy/controller.py

```diff
@@ -1160,41 +1139,50 @@ def set_last_wake_up_time(
             _LOGGER.debug("%s: Resetting last_wake_up_time to: %s", vin[-5:], timestamp)
             self._last_wake_up_time[vin] = timestamp
 
-    def set_car_online(
-        self, car_id: Text = None, vin: Text = None, online_status: bool = True
+    def _set_car_state(
+        self, new_state: str, car_id: Text = None, vin: Text = None
     ) -> None:
-        """Set online status for car_id.
+        """Set vehicle state for car_id or vin.
 
-        Will also update "last_wake_up_time" if the car changes from offline
+        Will also update "last_wake_up_time" if the car changes from asleep
         to online
 
         Parameters
         ----------
+        new_state: string
+            The state string from a vehicle api response
         car_id : string
             Identifier for the car on the owner-api endpoint.
         vin : string
             VIN number
 
-        online_status : boolean
-            True if the car is online (awake)
-            False if the car is offline (out of reach or sleeping)
-
-
         """
+        if not (car_id or vin):
+            raise ValueError("Either car_id or vin must be provided to set car state")
         if car_id and not vin:
             vin = self._id_to_vin(car_id)
-        if vin and self.get_car_online(vin=vin) != online_status:
+
+        if self.get_car_state(vin=vin) != new_state:
             _LOGGER.debug(
-                "%s: Changing car_online from %s to %s",
+                "%s: Changing car state from %s to %s",
                 vin[-5:],
-                self.get_car_online(vin=vin),
-                online_status,
+                self.get_car_state(vin=vin),
+                new_state,
             )
-            self.car_online[vin] = online_status
-            if online_status:
-                self.set_last_wake_up_time(vin=vin, timestamp=round(time.time()))
-
-    def get_car_online(self, car_id: Text = None, vin: Text = None):
+            self.__car_state[vin] = new_state
+            # Update the car object's state as well if it exists
+            if vin in self.cars:
+                self.cars[vin].update_car_info({"state": new_state, "vin": vin})
+
+        if self.is_car_asleep(vin=vin) and new_state == "online":
+            self.set_last_wake_up_time(vin=vin, timestamp=round(time.time()))
+        # Don't update __car_sleeping when state is offline or unavailable
+        if new_state == "asleep":
```

Comment:
We should probably turn these into const in const.py and use that so we don't have to search for them later if they change.

Reply:
good call, thanks